### PR TITLE
Conseil 168 Implementation of the openAPI doc using endpoints library

### PR DESCRIPTION
**Pull request by piotrkosecki**
contains implementation from previous two PRs:
https://github.com/Cryptonomic/Conseil/pull/252
https://github.com/Cryptonomic/Conseil/pull/265

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala [hunk 1]
```diff
@@ -39,27 +41,32 @@ object Conseil extends App with LazyLogging with EnableCORSDirectives with Conse
...
-                }
-              }
+              tezos.route ~
+              AppInfo.route
+            } ~
+            logRequest("Metadata Route", Logging.DebugLevel) {
+              platformDiscovery.route
+            } ~
+            logRequest("Data Route", Logging.DebugLevel) {
+              data.getRoutes ~ data.postRoute
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: `data.getRoute`? for more consistency... or is it already an api call?

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala [hunk 2]
```diff
@@ -8,13 +8,15 @@ import akka.http.scaladsl.server.Directives._
...
+import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport
 import tech.cryptonomic.conseil.config.ConseilAppConfig
 import tech.cryptonomic.conseil.directives.EnableCORSDirectives
 import tech.cryptonomic.conseil.routes._
+import tech.cryptonomic.conseil.routes.openapi.OpenApiDoc
 
 import scala.concurrent.ExecutionContextExecutor
 
-object Conseil extends App with LazyLogging with EnableCORSDirectives with ConseilAppConfig {
+object Conseil extends App with LazyLogging with EnableCORSDirectives with ConseilAppConfig with FailFastCirceSupport {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: request for info: how does `FailFast` work wrt to any other choice?

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala [hunk 3]
```diff
@@ -39,27 +41,32 @@ object Conseil extends App with LazyLogging with EnableCORSDirectives with Conse
...
+              data.getRoutes ~ data.postRoute
             }
-          } ~ options {
+          } ~
+          options {
             // Support for CORS pre-flight checks.
             complete("Supported methods : GET and POST.")
           }
         }
+      } ~
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: Can I ask for comments on matching braces when closing... I have the feeling there are mismatched braces

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala (hunk 2 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: there is two of them: `FailFast` and `ErrorAccumulating`, the difference is that the second one accumulates all of the errors during deserialization: https://github.com/hseeberger/akka-http-json/blob/master/akka-http-circe/src/main/scala/de/heikoseeberger/akkahttpcirce/CirceSupport.scala

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala (hunk 1 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: ok, renamed it

File: src/main/scala/tech/cryptonomic/conseil/Conseil.scala (hunk 3 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/Conseil.scala]: I don't think they are mismatched 😉

File: src/main/scala/tech/cryptonomic/conseil/routes/Data.scala [hunk 4]
```diff
@@ -26,102 +24,125 @@ object Data {
...
-            }
-          }
+  /** V2 Route implementation for query endpoint */
+  val postRoute: Route = queryEndpoint.implementedByAsync {
+    case ((platform, network, entity), apiQuery, _) =>
+      apiQuery.validate(entity).map { validQuery =>
+        platformNetworkValidation(platform, network) {
+          queryProtocolPlatform.queryWithPredicates(platform, entity, validQuery)
         }
+      }.left.map(Future.successful).bisequence.map(eitherOptionOps)
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: I'm a bit lost here, is it trying to expose errors as a result in a Future anyway, when there is something to show?

Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: yes, as validation of the query is synchronous and query result is a `Future`, but `implementedByAsync` needs `Future` so I had to do it this way

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataPlatform.scala [hunk 5]
```diff
@@ -25,7 +25,7 @@ class DataPlatform(operationsMap: Map[String, DataOperations]) {
     * @return query result as a option[map]
     * */
   def queryWithPredicates(platform: String, tableName: String, query: Query)
-    (implicit ec: ExecutionContext): Option[Future[List[Map[String, Option[Any]]]]] = {
+    (implicit ec: ExecutionContext): Option[Future[List[QueryResponse]]] = {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/generic/chain/DataPlatform.scala]: I guess we can improve consistency in the calls if we `sequence` the result here to return a
`Future[Option[List[QueryResponse]]]`

On a personal note, I find this more to my taste than the other way round

File: src/main/scala/tech/cryptonomic/conseil/routes/Data.scala [hunk 6]
```diff
@@ -26,102 +24,125 @@ object Data {
...
+    accountByIdRoute,
+    accountsRoute,
+    operationGroupByIdRoute,
+    operationGroupsRoute,
+    avgFeesRoute,
+    operationsRoute
+  )
+
+  /** Function for validation of the platform and network */
+  private def platformNetworkValidation[A](platform: String, network: String)(operation: Option[Future[A]]): Future[Option[A]] = {
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: also, we can get rid of this specific instance of validation if we refactor the `queryWithPredicates` return type to be consistent with old api (i.e. Future[Option[...]])

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package tech.cryptonomic.conseil.routes.openapi
+
+/** Helper trait which implements flatten for the tuples using shapeless library
+  * Taken from:
+  * https://gist.github.com/travisbrown/3945529
+  * */
+trait TupleFlattenHelper {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala]: I would love some help with documenting this

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/Validation.scala [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package tech.cryptonomic.conseil.routes.openapi
+
+import endpoints.algebra
+import endpoints.algebra.Documentation
+import tech.cryptonomic.conseil.generic.chain.DataTypes.QueryValidationError
+
+/** Trait adding validation for query  */
+trait Validation extends algebra.Responses {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/Validation.scala]: ok, done

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/QueryStringLists.scala [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package tech.cryptonomic.conseil.routes.openapi
+
+/** Trait adding parsing query strings as a lists */
+trait QueryStringLists extends endpoints.algebra.Urls {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/QueryStringLists.scala]: ok, done

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/QueryStringListsServer.scala [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package tech.cryptonomic.conseil.routes.openapi
+import akka.http.scaladsl.server.Directives._
+
+/** Trait providing query string list implementation for akka-http server */
+trait QueryStringListsServer extends QueryStringLists with endpoints.akkahttp.server.Endpoints {
+  /** Implementation of the query string list parsing for query params */
+  def qsList[A: QueryStringParam](name: String, docs: Option[String] = None): QueryString[List[A]] =
+    new QueryString[List[A]](parameter(name.as[A].*).tmap(_._1.toList))
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/QueryStringListsServer.scala]: my mistake, it was supposed to be map, changed

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/EndpointsHelpers.scala [hunk 11]
```diff
@@ -0,0 +1,61 @@
+package tech.cryptonomic.conseil.routes.openapi
+
+import cats.Functor
+import cats.syntax.functor._
+import endpoints.algebra
+import tech.cryptonomic.conseil.tezos.ApiOperations.Filter
+
+/** Trait containing helper functions which are necessary for parsing query parameter strings as Filter  */
+trait EndpointsHelpers extends QueryStringLists with algebra.JsonSchemaEntities with Validation with TupleFlattenHelper {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/EndpointsHelpers.scala]: added as self-type, renamed trait and qsFilter

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/EndpointsHelpers.scala [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      List[String],
+      List[String],
+      List[String],
+      List[String],
+      Option[String],
+      Option[String]
+    )
+
+  /** Function for extracting query string with query params */
+  def filterQs: QueryString[QueryParams] = {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/EndpointsHelpers.scala]: sure, done

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/OpenApiDoc.scala [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  with openapi.Endpoints
+  with openapi.JsonSchemaEntities
+  with openapi.BasicAuthentication {
+
+  /** OpenAPI JSON*/
+  def openapiJson: Json =
+    openapi.asJson
+
+  /** OpenAPI definition */
+  def openapi: OpenApi = openApi(Info("Conseil API", "0.0.1"))(queryEndpoint, blocksEndpoint, blocksHeadEndpoint, blockByHashEndpoint,
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/OpenApiDoc.scala]: no, changed to single column

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/DataEndpoints.scala [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+      tags = List("Query")
+    )
+
+  /** V2 Blocks endpoint definition */
+  def blocksEndpoint: Endpoint[((String, String, Filter), String), Option[List[QueryResponse]]] =
+    endpoint(
+      request = get(
+        url = commonPath / "blocks" /? queryStringFilter,
+        headers = header("apiKey")),
+      response = jsonResponse[List[QueryResponse]](docs = Some("Query compatibility endpoint for blocks")).orNotFound(Some("Not found")),
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/DataEndpoints.scala]: ok, used this function

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/OpenApiDoc.scala [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    accountsEndpoint, accountByIdEndpoint, operationGroupsEndpoint, operationGroupByIdEndpoint, avgFeesEndpoint, operationsEndpoint,
+    platformsEndpoint, networksEndpoint, entitiesEndpoint, attributesEndpoint, attributesValuesEndpoint, attributesValuesWithFilterEndpoint,
+    blocksEndpointV1, blocksHeadEndpointV1, blockByHashEndpointV1, accountsEndpointV1, accountByIdEndpointV1, operationGroupsEndpointV1,
+    operationGroupByIdEndpointV1, avgFeesEndpointV1, operationsEndpointV1, appInfoEndpoint
+  )
+
+  /** Function for validation definition in documentation */
+  def validated[A](response: List[OpenApiDoc.DocumentedResponse], invalidDocs: Documentation): List[OpenApiDoc.DocumentedResponse] =
+    response :+ OpenApiDoc.DocumentedResponse(
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/OpenApiDoc.scala]: ok, added more info in the doc

File: src/test/scala/tech/cryptonomic/conseil/routes/AppInfoRouteTest.scala [hunk 16]
```diff
@@ -15,10 +15,10 @@ class AppInfoRouteTest extends WordSpec with Matchers with ScalatestRouteTest {
     /* we need to embed in a complete route the fragment to test it
      * (see https://doc.akka.io/docs/akka-http/current/routing-dsl/testkit.html#testing-route-fragments)
      */
-    val sut = pathPrefix("info") { AppInfo.route }
+    val sut = AppInfo.route
 
     "expose an endpoint to get the current application version" in {
-      Get("/info") ~> sut ~> check {
+      Get("/info") ~> addHeader("apiKey", "hooman") ~> sut ~> check {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/routes/AppInfoRouteTest.scala]: 😉

File: src/main/scala/tech/cryptonomic/conseil/routes/Data.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: sure, changed

File: src/main/scala/tech/cryptonomic/conseil/routes/Data.scala [hunk 17]
```diff
@@ -26,102 +24,125 @@ object Data {
...
+    ConfigUtil.getNetworks(config, platform).find(_.network == network).flatMap { _ =>
+      operation
+    }.sequence
+  }
 
+  /** Function for validation of the platform and network with flatten */
+  private def platformNetworkValidationFlattened[A](platform: String, network: String)(operation: Future[Option[A]]): Future[Option[A]] = {
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: changed

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataPlatform.scala (hunk 5 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataPlatform.scala]: sure, changed

File: src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/routes/openapi/TupleFlattenHelper.scala]: also slightly refactored it - made changes suggested by the IDE, added types

**Pull request merged**